runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -1999,56 +2005,79 @@ public Iterable<Entry<K, V>> read() {
                     KeyState keyState = keyStateMap.get(structuralKey);
-                    if (!keyState.valuesCached) {
-                      keyState.values.extendWith(values);
-                      keyState.valuesCached = true;
-                    }
+                    keyState.existence = KeyExistence.KNOWN_EXIST;
+                    keyState.values.extendWith(entry.getValue());
+                    // We can't set keyState.valuesCached to true here, because there may be more
+                    // paginated values that should not be filtered out in filteredEntries.
                   });
               allKeysKnown = true;
               complete = true;
+              // Unload keys that are not known exist from cache, set valuesCached of all cached
+              // entries to true.
               keyStateMap
                   .entrySet()
                   .removeIf(
-                      entry ->
-                          entry.getValue().existence == KeyExistence.KNOWN_NONEXISTENT
-                              && !entry.getValue().removedLocally);
+                      entry -> {
+                        KeyState keyState = entry.getValue();
+                        boolean shouldRemove =
+                            (keyState.existence == KeyExistence.KNOWN_NONEXISTENT
+                                    && !keyState.removedLocally)
+                                || keyState.existence == KeyExistence.UNKNOWN_EXISTENCE;
+                        keyState.valuesCached = !shouldRemove;
+                        return shouldRemove;
+                      });
               return Iterables.unmodifiableIterable(mergedCachedEntries());
             } else {
-              MultimapIterables<K, V> local = mergedCachedEntries();
-              entryMap.forEach(
-                  (structuralKey, values) -> {
-                    KeyState keyState = keyStateMap.get(structuralKey);
-                    if (!keyState.valuesCached) {
-                      local.extendWith(keyState.originalKey, values);
-                    }
-                  });
-              return Iterables.unmodifiableIterable(local);
+              Iterable<Entry<K, V>> fromWindmill =
+                  Iterables.concat(
+                      Iterables.transform(
+                          Iterables.transform(
+                              filteredEntries,
+                              entry ->
+                                  new AbstractMap.SimpleEntry<>(
+                                      keyStateMap.get(entry.getKey()).originalKey,
+                                      entry.getValue())),
+                          entry ->
+                              Iterables.transform(
+                                  entry.getValue(),
+                                  v -> new AbstractMap.SimpleEntry<>(entry.getKey(), v))));
+              return Iterables.unmodifiableIterable(
+                  Iterables.concat(mergedCachedEntries(), fromWindmill));
```

Comment:
are the keys expected to be returned in order by this iterable? You might need to do some kind of merging iterable